### DISCUSSION TITLE
Any interest in a rich-text collaborative editing module?

### DISCUSSION THREAD

**Discussion by ProTip** [OrchardCMS/OrchardCore#15809] [General]
I created https://github.com/stepwisehq/prosemirror-dotnet to power the editor backend for the [StepWise](https://stepwisehq.com) MVP. ProseMirror is probably the most popular rich-text editor framework on the planet, so it's been surprisingly hard to get any feedback or acknowledgement on that project haha. ProseMirror is used at the NYT and powers Confluence for a couple examples.

Recently I have been considering the viability of extracting the document module from the StepWise project and creating a sort of drop-in library for use in other DotNet projects. But again, it's been hard getting any feedback on this idea.

Anyone have a sense on if a project like this would be in demand as a library or even an Orchard Core module?

**Comment by Piedone** [comment]
Ah interesting, I'm glad you found Orchard Core!

For editing human text we currently use Trumbowyg for HTML and EasyMDE for Markdown (for editing Liquid and JS code we use CodeMirror). It's specifically supported to add your own editor, so it's possible to add a ProseMirror.Net HTML editor, for example, and use that to create pages or whatever else.

A simple editor integration like this is simple enough, but ideally, one would also integrate the Media Library browser that displays the Media Library (with images and else) as a popup, and lets you e.g. insert images into a text.

The question is, what advantages would this bring over Trumbowyg? Unfortunately, I don't know ProseMirror, so I can't comment.

**Comment by ProTip** [comment]
I can't speak to Trumbowyg as I haven't used it, but it doesn't look like it supports collaborative editing?

ProseMirror supports building real-time collaborative editing experiences such as found in Google Docs(ignoring they moved to canvas), Confluence, and Microsoft Loop.

The functionality I'm thinking about extracting from StepWise is the backend functionality to facilitate this. It's the central authority that accepts individual document edits, authorizes them, validates them against the document schema, stores them, and broadcasts them out to the other connected clients. It tracks all edits and supports history, versioning, reverting, publishing, and more.

**Reply by Piedone** [reply]
Ah I see, so that's what you mean by "collaborative". That would actually be awesome, yes!

The challenge I see is around storage: OC uses its own storage implementation, of course, so this needs to somehow work with it. Perhaps the easiest would be for you to make the component reusable everywhere if you'd use some transient server-side storage (like memory cache), not to have any further integration with databases. That would also nicely fit into what OC does, since then the feature could be just a compartmentalized editor that would work on its own. Then, once someone saves the content item, the data would be stored in the OC database as usual.

**Reply by Piedone** [reply]
There's actually an ancient issue for this :) https://github.com/OrchardCMS/OrchardCore/issues/1601.

**Comment by giannik** [comment]
@ProTip  I find this interesting for LineOfBusiness cases . most of the wysiwyg editors offer mostly the same things . 
I would suggest creating a separate module similar to : https://github.com/emrahtokalak/Emrah.OrchardCore.CKEditor that did something similar for ck editor.
